File: backend/app/models/impact.py
from pydantic import BaseModel, ConfigDict


class ImpactModel(BaseModel):
    model_config = ConfigDict(populate_by_name=True)


class NearEarthObject(ImpactModel):
    id: str
    name: str
    nasaJplUrl: str | None = None
    estimatedDiameterMinKm: float | None = None
    estimatedDiameterMaxKm: float | None = None
    potentiallyHazardous: bool
    closeApproachDate: str
    closeApproachDateTime: str | None = None
    relativeVelocityKmS: float | None = None
    missDistanceKm: float | None = None
    missDistanceLunar: float | None = None
    orbitingBody: str | None = None


class Fireball(ImpactModel):
    id: str
    dateTime: str
    latitude: float | None = None
    longitude: float | None = None
    altitudeKm: float | None = None
    velocityKmS: float | None = None
    energy: float
    impactEnergyKt: float
    locationDescription: str | None = None

File: frontend/src/types/impact.ts
export type NearEarthObject = {
  id: string;
  name: string;
  nasaJplUrl: string | null;
  estimatedDiameterMinKm: number | null;
  estimatedDiameterMaxKm: number | null;
  potentiallyHazardous: boolean;
  closeApproachDate: string;
  closeApproachDateTime: string | null;
  relativeVelocityKmS: number | null;
  missDistanceKm: number | null;
  missDistanceLunar: number | null;
  orbitingBody: string | null;
};

export type Fireball = {
  id: string;
  dateTime: string;
  latitude: number | null;
  longitude: number | null;
  altitudeKm: number | null;
  velocityKmS: number | null;
  energy: number;
  impactEnergyKt: number;
  locationDescription: string | null;
};

export type ImpactResponse<T extends 'neos' | 'fireballs'> = {
  total: number;
  updatedAt: string;
  source: 'live-or-cache' | 'unavailable';
  provider: string;
  error: { code: string; message: string } | null;
} & Record<T, T extends 'neos' ? NearEarthObject[] : Fireball[]>;
